### PR TITLE
Add specs for a safe navigator.

### PR DESCRIPTION
**Pull request by iliabylich**

### CODE REVIEW

**Comment by eregon** [comment]
The build failed because the parser itself fails on < 2.3.
The usual trick is to wrap things in eval, so older versions can still require the file.
We can either wrap every `&.` code in eval or have a big eval inside the version guard with a heredoc string.

**Comment by iliabylich** [comment]
Updated PR to use `eval("").should` :smile:

**Comment by iliabylich** [comment]
@eregon by the way, did I miss any cases?

**Comment by elia** [comment]
@iliabylich maybe stuff like `nil&.[](123)&.asdfdas` / `nil&.[](123) {|a| foobar }&.asdfdas`?

**Comment by iliabylich** [comment]
@elia this PR has a test for a block and method arguments to be passed. Or do you mean a test for all these stuff on `nil`?

From https://github.com/ruby/spec/issues/175:
```
+ method name is syntactically required obj.try! {} # valid obj&. {} # syntax error
+ arguments are evaluated only if a call is made: obj.try!(:foo, bar()) # bar() is always evaluated obj&.foo(bar()) # bar() is conditionally evaluated
+ attribute assignment is valid obj&.attr += 1 
```

All of them are not tests here :smile: . @eregon Don't merge this PR please, I'm going to update it

**Comment by eregon** [comment]
@iliabylich I was going to point you to https://github.com/ruby/spec/issues/175 indeed :)

**Comment by elia** [comment]
@iliabylich the idea was about having an example that combined a bunch of cases together to ensure that's supported as well

**Comment by iliabylich** [comment]
@eregon Done I guess

@elia [this] (https://github.com/iliabylich/rubyspec/blob/ffabc7f732b3215a271079482c8cadebc7184d74/language/safe_navigator_spec.rb#L19-L24) and [this](https://github.com/iliabylich/rubyspec/blob/ffabc7f732b3215a271079482c8cadebc7184d74/language/safe_navigator_spec.rb#L15-L17) tests do it

File: language/safe_navigator_spec.rb [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+        def initialize
+          @m = 0
+        end
+      end
+
+      obj = klass.new
+
+      eval("obj&.m += 3")
+      obj.m.should == 3
```
Comment (eregon) [inline language/safe_navigator_spec.rb]: Maybe just one last case: attr assignment with nil:
```ruby
obj = nil
eval("obj&.m += 3").should == nil
```

Comment (eregon) [inline language/safe_navigator_spec.rb]: I'll add it myself, too easy to fix :)

**Comment by eregon** [comment]
@iliabylich Thanks a lot for the nice new specs!

File: language/safe_navigator_spec.rb (hunk 1 above)
Comment (eregon) [inline language/safe_navigator_spec.rb]: (https://github.com/ruby/spec/commit/be3923620b6f309e181c8665d14b1200e16e9255)

**Pull request merged**